### PR TITLE
Automated Device Configuration with Event-Driven Filters

### PR DESCRIPTION
**Pull request by Danielius1922**
## Description

This feature allows users to pre-configure devices by specifying desired configurations that are automatically applied when a ResourceChanged event occur. The newly introduced `snippet service` is responsible for managing this process. 

#### Key Features:
1. **Device Configuration**:
   - Users can create and store configuration profiles for devices.
   - These configurations are automatically applied when a device connects or a ResourceChanged event occurs.

2. **Event-Driven Configuration**:
   - The configuration is triggered by the `ResourceChanged` event.
   - Users can set filters to define the conditions under which configurations are applied.

3. **Filter Conditions**:
   - Users must set at least one of the following filters:
     - **DeviceID filter**: Trigger based on specific device IDs.
     - **Resource-Type filter**: Trigger when all specified resource types are present.
     - **href filter**: Trigger based on specific href values.
   - All specified filters must be met for the configuration to be applied.
   - For DeviceID and href filters, any listed value can trigger the filter.
   - For the Resource-Type filter, all listed values must be present.

4. **Optional JQ Expression**:
   - Users can optionally use a JQ expression to filter the content of the `ResourceChanged` event.
   - The JQ expression must evaluate to true for the configuration to be applied.

5. **User Interface**:
   - A user-friendly UI is provided to manage configurations and set filters easily.

6. **APIs**:
   - **gRPC API**: For detailed API operations, refer to the [gRPC API documentation](https://github.com/plgd-dev/hub/blob/main/snippet-service/pb/service.proto).
   - **HTTP API**: For HTTP API operations, refer to the [HTTP API documentation](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/plgd-dev/hub/main/snippet-service/pb/service.swagger.json).

#### Benefits:
- **Automated Device Management**: Simplifies the process of configuring devices by automating the application of configurations based on specific events and conditions.
- **Flexible Filtering**: Provides robust filtering options to ensure configurations are applied only under desired conditions.
- **Enhanced Control**: Users have greater control over device configurations and can tailor them to meet specific needs and scenarios.

This new feature enhances the capability of the system to manage and configure devices efficiently, ensuring they are always set up as required for optimal operation.

### CODE REVIEW

**Comment by Danielius1922** [comment]
@coderabbitai review

File: snippet-service/test/configuration.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+			configurations[conf.GetId()] = configuration
+		}
+		configuration.Versions = append(configuration.Versions, store.ConfigurationVersion{
+			Version:   conf.GetVersion(),
+			Resources: conf.GetResources(),
+		})
+		configurations[conf.GetId()] = configuration
+	}
+	return configurations
+}
```
[the remark this answers is not in the document]
Comment (Danielius1922) [reply snippet-service/test/configuration.go]: code in `test` folder is testing code, we are less strict about testing code, lets ignore function complexity for testing code

File: snippet-service/config.yaml
Comment (jkralik) [inline snippet-service/config.yaml:1]: TODO: remove values which are not used

File: .vscode/settings.json [hunk 2]
```diff
@@ -34,6 +34,8 @@
         "TEST_GRPC_GATEWAY_LOG_DUMP_BODY": "false",
         "TEST_IDENTITY_STORE_LOG_LEVEL": "info",
         "TEST_IDENTITY_STORE_LOG_DUMP_BODY": "false",
+        "TEST_SNIPPET_SERVICE_LOG_LEVEL": "debug",
```
Comment (Danielius1922) [inline .vscode/settings.json]: info

File: charts/plgd-hub/templates/snippet-service/_helpers.tpl [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+{{- end }}
+{{- end }}
+
+{{- define  "plgd-hub.snippetservice.image" -}}
+    {{- $registryName := .Values.snippetservice.image.registry | default "" -}}
+    {{- $repositoryName := .Values.snippetservice.image.repository -}}
+    {{- $tag := .Values.snippetservice.image.tag | default .Chart.AppVersion | toString -}}
+    {{- printf "%s%s:%s" $registryName $repositoryName  $tag -}}
+{{- end -}}
+
```
Comment (jkralik) [inline charts/plgd-hub/templates/snippet-service/_helpers.tpl]: Replace with `plgd-hub.image` - look to another services how it is used.

File: charts/plgd-hub/values.yaml [hunk 4]
```diff
@@ -2245,6 +2246,284 @@ certificateauthority:
     validFrom: "now-1h"
     expiresIn: "87600h"
 
+snippetservice:
+  # -- Enable certificate-authority service
```
Comment (jkralik) [inline charts/plgd-hub/values.yaml]: replace `certificate-authority` to `snippet-service` in comments

File: charts/plgd-hub/values.yaml [hunk 5]
```diff
@@ -2292,6 +2293,284 @@ certificateauthority:
...
+          keyFile:
+          certFile:
+          useSystemCAPool: false
+        bulkWrite:
+          # -- A time limit for write bulk to mongodb. A Timeout of zero means no timeout.
+          timeout: 1m0s
+          # -- The amount of time to wait until a record is written to mongodb. Any records collected during the throttle time will also be written. A throttle time of zero writes immediately. If recordLimit is reached, all records are written immediately
+          throttleTime: 500ms
+          # -- The maximum number of documents to cache before an immediate write.
+          documentLimit: 1000
```
Comment (jkralik) [inline charts/plgd-hub/values.yaml]: Is it used this part?

File: charts/plgd-hub/values.yaml [hunk 6]
```diff
@@ -2292,6 +2293,284 @@ certificateauthority:
...
+        keyspace:
+          name: plgdhub
+          create: true
+          replication:
+            class: SimpleStrategy
+            replication_factor: 1
+        tls:
+          caPool:
+          keyFile:
+          certFile:
```
Comment (jkralik) [inline charts/plgd-hub/values.yaml]: Do we support cqldb -> If not we need to remove this part for now

**Comment by Danielius1922** [comment]
@coderabbitai write summary for the PR, focus on the new snippet-service, you can ignore the rest

**Comment by Danielius1922** [comment]
@coderabbitai thanks for the summary, now write a description focusing on the snippet-service, that can be used as a commit message

File: charts/plgd-hub/values.yaml (hunk 5 above)
Comment (Danielius1922) [reply charts/plgd-hub/values.yaml]: no bulkwrite is not used currently in the snippet-service

File: charts/plgd-hub/values.yaml (hunk 6 above)
Comment (Danielius1922) [reply charts/plgd-hub/values.yaml]: the configuration exists, but the implementation for cqldb are just empty functions that return error

File: snippet-service/store/condition.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+		DeviceIdFilter:     c.Latest.DeviceIdFilter,
+		ResourceTypeFilter: c.Latest.ResourceTypeFilter,
+		ResourceHrefFilter: c.Latest.ResourceHrefFilter,
+		JqExpressionFilter: c.Latest.JqExpressionFilter,
+		ApiAccessToken:     c.Latest.ApiAccessToken,
+	}, nil
+}
+
+func (c *Condition) GetCondition(index int) *pb.Condition {
+	return &pb.Condition{
```
Comment (jkralik) [inline snippet-service/store/condition.go]: check index?

File: snippet-service/store/configuration.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+		Owner:     c.Owner,
+		Version:   c.Latest.Version,
+		Name:      c.Latest.Name,
+		Resources: c.Latest.Resources,
+		Timestamp: c.Latest.Timestamp,
+	}, nil
+}
+
+func (c *Configuration) GetConfiguration(index int) *pb.Configuration {
+	return &pb.Configuration{
```
Comment (jkralik) [inline snippet-service/store/configuration.go]: check index

File: charts/plgd-hub/values.yaml (hunk 5 above)
Comment (Danielius1922) [reply charts/plgd-hub/values.yaml]: removed

File: charts/plgd-hub/values.yaml (hunk 6 above)
Comment (Danielius1922) [reply charts/plgd-hub/values.yaml]: removed

**Review by jkralik — APPROVED** [review]

**Comment by Danielius1922** [comment]
@coderabbitai the code review was skipped due to file limit, but can you do code review if you examine only files in the snippet-service folder? or at least write description for the code in the snippet-service folder?

**Comment by Danielius1922** [comment]
@coderabbitai that's too long for a commit message, can you write something better? an overview of the code in the snippet-service folder that can be used as a commit message.

**Comment by Danielius1922** [comment]
@coderabbitai write a description of the code in the snippet-service folder, focusing on the overview of functionality of the added snippet-service

**Pull request merged**